### DISCUSSION TITLE
WinUI SizeToContent

### DISCUSSION THREAD

**Discussion by gbohrn** [microsoft/microsoft-ui-xaml#9404] [Q&A] [answered]
I am trying to make a simple dialog app for desktop alerting.  This was quite easy in WPF as I could use SizeToContent to have the app resize itself to the content generated.  I have not found a way to do this as WinUI app does not resize to the content in the Main Window.  I chose to try to move to WinUI because a self contained WPF app is ginormous.  WinUI is still large but a fraction of the size of a WPF app.
I understand I can use the AppWindow to resize the app, but as I do not know the size of the content before the window starts rendering, I cannot set the size.  As mentioned, WPF SizeToContent makes this easy.  Could use some hlep.

**Comment by powext** [comment]
Up. Did you find a solution?

**Reply by gbohrn** [reply]
I have not.  There currently does not appear to be any solid solution out there that replicates this basic functionality in WPF.

**Reply by powext** [reply]
Currently my raw workaround is to get the actual size of the content and change the window size right after it becomes loaded

**Reply by gbohrn** [reply]
I have tried this.  It doesn't work well for my situation.  I am implementing a basic alerting app that has different sizes (different user controls swapped during startup).  I do not know the size of the user control until the app is already starting to render.  I tried setting hard code based on the dialog, but then it needs to fluctuate based on localization.  So, until I can get this functionality reliably I still have to deploy via WPF.  this is painful as my app needs to be self contained and WinUI self contained is about 1/10 of the size of the WPF app.  I am deploying this on the fly from cloud services to millions of endpoints, so needless to say the bandwidth usage is brutal (and the cloud expenses are not cheap).  This should have been basic functionality for a desktop app.

**Reply by pcruche** [reply]
What do you mean with 'doesn't work well'? I currently have a situation that seems to be similar, where I'm loading the actual content configuration for a window from runtime data.
I use this to configure the root container of the window page (a stack panel), and I'm doing this immediately before activation of the window. Once configured, I define the root container as Content of the page and add a handler for the panel's Loaded-event. In this handler I can resize the window based on the panel's ActualWidth and ActualHeight properties. That initial resizing is visible, but I found it to be good enough for my use case.

**Reply by gbohrn** [reply]
I will need to reread through what you wrote and see.  My simple dialog app has 4 user controls that are loaded (only 1) based on commandline input.  This includes any localization strings.  The user controls do not have a set size because of localization so I do not have their size until prerender.  There are currently 4 different usercontrol options so far but expecting to add more.

I am thinking what may make sense is I just turn each into a window instead of using a Window as the shell to host different usercontrols which is what I am currently doing.  Previously, the outer window renders before the usercontrol dimensions are calculated so I am unable to reset the window size based on the user control.

It would be nice if Microsoft just added this to the window class as they have in WPF which makes this stuff trivial.

**Comment by yourordinarycat** [comment]
Add a `SizeChanged` event handler to your Window's root element. The event handler should look roughly like this:

```C#
private void OnRootSizeChanged(object sender, SizeChangedEventArgs e)
{
    var elm = (FrameworkElement)sender;
    elm.SizeChanged -= OnRootSizeChanged;

    // Edit from original: account for scaling
    var scale = elm.XamlRoot.RasterizationScale;
    var height = (int)Math.Ceiling(elm.DesiredSize.Height * scale);
    var width = (int)Math.Ceiling(elm.DesiredSize.Width * scale);

    AppWindow.ResizeClient(new(width, height));
    Activate();
}
```

This should be a good replacement, unless I'm missing something.

**Reply by gbohrn** [reply]
This is helpful, but unfortunately the app renders a much larger blank window before the sizechanged occurs.  I have not found a way to prevent the UI frame from showing before the maingrid is rendered.  So the app takes 3/4 of the screen when I basically need a simple dialog app.  This works perfectly fine in WPF with SizeToContent.  This is glaringly missing in WinUI and is starting to make this framework frankly unusable.

**Reply by yourordinarycat** [reply, marked as answer]
> app renders a much larger blank window before the sizechanged occurs. Probably should not show the ui until the sizechanged occurs

@gbohrn I think not calling `Activate` before the initial `SizeChanged` should suffice.

**Reply by gbohrn** [reply]
Thanks, that was the missing part.  I moved activate to the onLoaded of my main Grid.  Fixed it.

**Reply by domenic** [reply]
This worked OK for me, but in my app it fired 3 times during startup.

I ended up just using the debugger to record the 3rd resize and then hard-coding those values.
